extension/persistence/relational-jdbc/src/main/java/org/apache/polaris/extension/persistence/relational/jdbc/DatasourceOperations.java

```diff
@@ -0,0 +1,174 @@
+              LOGGER.debug("Query {} executed {} rows affected", sql, rowsUpdated);
+            } catch (SQLException e) {
+              LOGGER.error("Error executing query {}", sql, e);
+              // re:throw this as unhandled exception
+              throw new RuntimeException(e);
+            }
+            sqlBuffer.setLength(0); // Clear the buffer for the next statement
+          }
+        }
+      }
+      connection.setAutoCommit(autoCommit);
+    } catch (IOException e) {
+      LOGGER.error("Error reading the script file", e);
+      throw new RuntimeException(e);
+    } catch (SQLException e) {
+      LOGGER.error("Error executing the script file", e);
+      throw e;
+    }
+  }
+
+  public <T, R> List<R> executeSelect(
+      @Nonnull String query,
+      @Nonnull Class<T> targetClass,
+      @Nonnull Function<T, R> transformer,
+      Predicate<R> entityFilter,
+      int limit)
+      throws SQLException {
+    try (Connection connection = borrowConnection();
+        Statement statement = connection.createStatement();
+        ResultSet s = statement.executeQuery(query)) {
+      return ResultSetToObjectConverter.collect(s, targetClass, transformer, entityFilter, limit);
+    } catch (SQLException e) {
+      LOGGER.error("Error executing query {}", query, e);
+      throw e;
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  public int executeUpdate(String query) throws SQLException {
+    try (Connection connection = borrowConnection();
+        Statement statement = connection.createStatement()) {
+      boolean autoCommit = connection.getAutoCommit();
+      connection.setAutoCommit(true);
+      int result = statement.executeUpdate(query);
+      connection.setAutoCommit(autoCommit);
+      return result;
+    } catch (SQLException e) {
+      LOGGER.error("Error executing query {}", query, e);
+      throw e;
+    }
+  }
+
+  public int executeUpdate(String query, Statement statement) throws SQLException {
+    LOGGER.debug("Executing query {} within transaction", query);
+    try {
+      return statement.executeUpdate(query);
+    } catch (SQLException e) {
+      LOGGER.error("Error executing query {}", query, e);
+      throw e;
+    }
+  }
+
+  public void runWithinTransaction(TransactionCallback callback) throws SQLException {
+    try (Connection connection = borrowConnection()) {
+      boolean autoCommit = connection.getAutoCommit();
+      connection.setAutoCommit(false);
+      boolean success = false;
+      try {
+        try (Statement statement = connection.createStatement()) {
+          success = callback.execute(statement);
+        }
+      } finally {
+        if (success) {
+          connection.commit();
+        } else {
+          connection.rollback();
+        }
+        connection.setAutoCommit(autoCommit);
+      }
+    } catch (SQLException e) {
+      LOGGER.error("Caught Error while executing transaction", e);
```

Comment:
I'm not sure this is justified. Failed Tx may be retried at higher levels and these error could cause false alarms.

If you're concerned with losing exceptions, let's log as "debug".

Reply:
I see a "thumbs up" reaction on my comment, but the code is unchanged :sweat_smile: Do you prefer to keep this as an ERROR log?

Same concern about lines 160 an d138.

Reply:
I did address this feeback, but later feeback was to make it back to error : https://github.com/apache/polaris/pull/1287#discussion_r2051583483

Hence changes please let me know, what do you think considering this.

Reply:
from my POV since the exception is re-thrown, ERROR log is not justified - by re-throwing we're delegating error handling to the caller. If the caller is able to recover from the error, then the ERROR log is a false alarm.